mvebu: add Globalscale MOCHAbin

robimarko opened the pull request:
Globalscale MOCHAbin is a Armada 7040 based development board.

Specifications:
* Armada 7040 Quad core ARMv8 Cortex A-72 @ 1.4GHz
* 2 / 4 / 8 GB of DDR4 DRAM
* 16 GB eMMC
* 4MB SPI-NOR (Bootloader)
* 1x M.2-2280 B-key socket (for SSD expansion, SATA3 only)
* 1x M.2-2250 B-key socket (for modems, USB2.0 and I2C only)
* 1x Mini-PCIe 3.0 (x1, USB2.0 and I2C)
* 1x SATA 7+15 socket (SATA3)
* 1x 16-pin (2×8) MikroBus Connector
* 1x SIM card slot (Connected to the mini-PCIe and both M.2 slots)
* 2x USB3.0 Type-A ports via SMSC USB5434B hub
* Cortex 2x5 JTAG
* microUSB port for UART (PL2303GL/PL2303SA onboard)
* 1x 10G SFP+
* 1x 1G SFP (Connected to 88E1512 PHY)
* 1x 1G RJ45 with PoE PD (Connected to 88E1512 PHY)
* 4x 1G RJ45 ports via Topaz 88E6141 switch
* RTC with battery holder (SoC provided, requires CR2032 battery)
* 1x 12V DC IN
* 1x Power switch
* 1x 12V fan header (3-pin, power only)
* 1x mini-PCIe LED header (2x0.1" pins)
* 1x M.2-2280 LED header (2x0.1" pins)
* 6x Bootstrap jumpers
* 1x Power LED (Green)
* 3x Tri-color RGB LEDs (Controllable)
* 1x Microchip ATECC608B secure element

Note that 1G SFP and 1G WAN cannot be used at the same time as they are in
parallel connected to the same PHY.

Installation:

Copy dtb from build_dir to bin/ and run tftpserver there:
```
$ cp ./build_dir/target-aarch64_cortex-a72_musl/linux-mvebu_cortexa72/image-armada-7040-mochabin.dtb bin/targets/mvebu/cortexa72/
$ in.tftpd -L -s bin/targets/mvebu/cortexa72/
```

Connect to the device UART via microUSB port and power on the device.

Power on the device and hit any key to stop the autoboot.

Set serverip (host IP) and ipaddr (any free IP address on the same subnet), e.g:
```
$ setenv serverip 192.168.1.10 # Host
$ setenv ipaddr 192.168.1.15 # Device
```

Set the ethernet device (Example for the 1G WAN):
`$ setenv ethact mvpp2-2`

Ping server to confirm network is working:
```
$ ping $serverip
Using mvpp2-2 device
host 192.168.1.15 is alive
```

Tftpboot the firmware:
```
$ tftpboot $kernel_addr_r openwrt-mvebu-cortexa72-globalscale_mochabin-initramfs-kernel.bin
$ tftpboot $fdt_addr_r image-armada-7040-mochabin.dtb
```

Boot the image:
`$ booti $kernel_addr_r - $fdt_addr_r`

Once the initramfs is booted, transfer `openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img.gz`
to /tmp dir on the device.

Gunzip and dd the image:
```
$ gunzip /tmp/openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img.gz
$ dd if=/tmp/openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img of=/dev/mmcblk0 && sync
```

Reboot the device.

Hit any key to stop the autoboot.

Reset U-boot env and set the bootcmd:
```
$ env default -a
$ setenv bootcmd 'load mmc 0 ${loadaddr} boot.scr && source ${loadaddr}'
```

Optionally I would advise to edit the console env variable to remove earlycon as that
causes the kernel to never use the driver for the serial console.
Earlycon should be used only for debugging before the kernel can configure the console
and will otherwise cause various issues with the console.

`$ setenv console 'console=ttyS0,115200'`

Save and reset
```
$ saveenv
$ reset
```

OpenWrt should boot from eMMC now.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

neheb commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-7040-mochabin.dts:
```diff
@@ -0,0 +1,448 @@
+
+			swport1: port@1 {
+				reg = <1>;
+				label = "lan0";
```
lan0?

robimarko replied:
Sorry, I don't get the issue?

neheb replied:
First time I see that. Most routers start lan numbers with 1.

robimarko replied:
Haven't really thought about that.
I would prefer to leave it at 0 so I don't have to send a new Linux DTS version upstream.

hauke replied:
Normally this should match the labels which are printed on the casing or used in the manual. If there are no labels defined this should be ok.

robimarko replied:
Yeah, I agree however there is nothing printed on the case and the case is optional.
The quick start guide I have does use lan0 as well.

erdoukki replied:
lan0 is used for first ports in all GlobalScale MVEBU boards...
EspressoBin V5 and V7 and EspressoBin-Ultra have LAN ports starting at 0 !

adschm commented on target/linux/mvebu/image/cortexa72.mk:
```diff
@@ -1,3 +1,11 @@
+  $(call Device/Default-arm64)
+  DEVICE_VENDOR := Globalscale
+  DEVICE_MODEL := MOCHAbin
+  DEVICE_DTS := armada-7040-mochabin
```
You should be able to use "SOC := armada-7040" instead

robimarko replied:
Will update it

robimarko replied:
Done

hauke replied:
I do not see the change

robimarko replied:
Sorry, pushed to the wrong branch.
Its pushed correctly now.

robimarko committed:
mvebu: add Globalscale MOCHAbin

Globalscale MOCHAbin is a Armada 7040 based development board.

Specifications:
* Armada 7040 Quad core ARMv8 Cortex A-72 @ 1.4GHz
* 2 / 4 / 8 GB of DDR4 DRAM
* 16 GB eMMC
* 4MB SPI-NOR (Bootloader)
* 1x M.2-2280 B-key socket (for SSD expansion, SATA3 only)
* 1x M.2-2250 B-key socket (for modems, USB2.0 and I2C only)
* 1x Mini-PCIe 3.0 (x1, USB2.0 and I2C)
* 1x SATA 7+15 socket (SATA3)
* 1x 16-pin (2×8) MikroBus Connector
* 1x SIM card slot (Connected to the mini-PCIe and both M.2 slots)
* 2x USB3.0 Type-A ports via SMSC USB5434B hub
* Cortex 2x5 JTAG
* microUSB port for UART (PL2303GL/PL2303SA onboard)
* 1x 10G SFP+
* 1x 1G SFP (Connected to 88E1512 PHY)
* 1x 1G RJ45 with PoE PD (Connected to 88E1512 PHY)
* 4x 1G RJ45 ports via Topaz 88E6141 switch
* RTC with battery holder (SoC provided, requires CR2032 battery)
* 1x 12V DC IN
* 1x Power switch
* 1x 12V fan header (3-pin, power only)
* 1x mini-PCIe LED header (2x0.1" pins)
* 1x M.2-2280 LED header (2x0.1" pins)
* 6x Bootstrap jumpers
* 1x Power LED (Green)
* 3x Tri-color RGB LEDs (Controllable)
* 1x Microchip ATECC608B secure element

Note that 1G SFP and 1G WAN cannot be used at the same time as they are in
parallel connected to the same PHY.

Installation:

Copy dtb from build_dir to bin/ and run tftpserver there:
$ cp ./build_dir/target-aarch64_cortex-a72_musl/linux-mvebu_cortexa72/image-armada-7040-mochabin.dtb bin/targets/mvebu/cortexa72/
$ in.tftpd -L -s bin/targets/mvebu/cortexa72/

Connect to the device UART via microUSB port and power on the device.

Power on the device and hit any key to stop the autoboot.

Set serverip (host IP) and ipaddr (any free IP address on the same subnet), e.g:
$ setenv serverip 192.168.1.10 # Host
$ setenv ipaddr 192.168.1.15 # Device

Set the ethernet device (Example for the 1G WAN):
$ setenv ethact mvpp2-2

Ping server to confirm network is working:
$ ping $serverip
Using mvpp2-2 device
host 192.168.1.15 is alive

Tftpboot the firmware:
$ tftpboot $kernel_addr_r openwrt-mvebu-cortexa72-globalscale_mochabin-initramfs-kernel.bin
$ tftpboot $fdt_addr_r image-armada-7040-mochabin.dtb

Boot the image:
$ booti $kernel_addr_r - $fdt_addr_r

Once the initramfs is booted, transfer openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img.gz
to /tmp dir on the device.

Gunzip and dd the image:
$ gunzip /tmp/openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img.gz
$ dd if=/tmp/openwrt-mvebu-cortexa72-globalscale_mochabin-squashfs-sdcard.img of=/dev/mmcblk0 && sync

Reboot the device.

Hit any key to stop the autoboot.

Reset U-boot env and set the bootcmd:
$ env default -a
$ setenv bootcmd 'load mmc 0 ${loadaddr} boot.scr && source ${loadaddr}'

Optionally I would advise to edit the console env variable to remove earlycon as that
causes the kernel to never use the driver for the serial console.
Earlycon should be used only for debugging before the kernel can configure the console
and will otherwise cause various issues with the console.

$ setenv console 'console=ttyS0,115200'

Save and reset
$ saveenv
$ reset

OpenWrt should boot from eMMC now.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

erdoukki commented:
@robimarko Depending of uboot commands available, the flashing may be simplified ?
Look at : https://github.com/erdoukki/vrac/tree/master/openwrt for my own method tested on EspressoBin and EspressoBinUltra !

robimarko commented:
I would prefer to not rely on the U-boot tools for flashing, especially since I dont know with what commands and version will it ship.
Especially since upstream ATF and U-boot is being worked on, this is an agnostic way of flashing.
And I have seen way too many bricked boards due to using U-boot tools without understanding how they operate.

erdoukki commented:
> I would prefer to not rely on the U-boot tools for flashing, especially since I dont know with what commands and version will it ship.

You're right, I will add an advice on my tutorial !

hauke commented:
Thank you for the patch, I applied it to master. 
Could you please send an update when your patch was accepted upstream. If it is upstream it would probably be easier to maintain this as a patch and not as a file.

adschm commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-7040-mochabin.dts:
```diff
@@ -0,0 +1,448 @@
+		};
+
+		led1_green: led@2 {
+			label = "led1:green";
```
Just saw that all labels are inverted to what I would expect: color:function, i.e. green:led1 etc. in this case.

Has this file already been upstreamed/submitted in kernel? If yes, this should be changed there. If no, I'd just change it locally myself?

robimarko replied:
No, it's got a review, but it's not picked yet.
I will change the DTS and send the new revision upstream.
I might as well add the color definitions as well at the same time.

Sorry for forgetting this.

adschm replied:
Be aware that we don't support the color/function _properties_ if you want to use led-* aliases. But since this is not the case at the moment, you might as well change.
I'd not except the kernel to accept color properties and label at the same time, though.

robimarko replied:
Yeah, predefined functions don't really apply here.
I usually define the colors in the DTS as well, which makes it easy to parse them although it has no functional difference.

robimarko replied:
https://patchwork.ozlabs.org/project/openwrt/patch/20211002175757.1019594-1-robert.marko@sartura.hr/

robimarko replied:
@adschm Can you pick the LED patch?
https://patchwork.ozlabs.org/project/openwrt/patch/20211002175757.1019594-1-robert.marko@sartura.hr/

robimarko commented:
@hauke Sure, I was planing to make it a backport when it got merged as the only incompatible thing should be the UTMI node as it only got added after 5.10 along with the UTMI driver.
